packages/effect/src/internal/stream.ts

```diff
@@ -5709,42 +5710,49 @@ export const shareRefCount = dual<
     readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
   }) => <R>(
     self: Stream.Stream<A, E, R>
-  ) => Stream.Stream<A, E, R>,
+  ) => Effect.Effect<Stream.Stream<A, E, R>, never, R | Scope.Scope>,
   <A, E, R>(
     self: Stream.Stream<A, E, R>,
     config: {
       readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
     }
-  ) => Stream.Stream<A, E, R>
+  ) => Effect.Effect<Stream.Stream<A, E, R>, never, R | Scope.Scope>
 >(
   2,
   <A, E, R>(
     self: Stream.Stream<A, E, R>,
     config: {
       readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
     }
-  ): Stream.Stream<A, E, R> => {
-    let refCount = 0
-    let connector: PubSub.PubSub<Take.Take<A, E>> | null = null
-    let fiber: Fiber.RuntimeFiber<void> | null = null
+  ): Effect.Effect<Stream.Stream<A, E, R>, never, R | Scope.Scope> => {
     return Effect.gen(function*() {
-      refCount++
-      connector ??= yield* (config.connector ?? PubSub.bounded<Take.Take<A, E>>(16))
-      fiber ??= yield* Effect.forkDaemon(runIntoPubSub(self, connector))
-      return flattenTake(fromPubSub(connector))
-    }).pipe(
-      unwrap,
-      ensuring(
-        Effect.suspend(() => {
-          refCount--
-          const cleanup = fiber
-          fiber = null
-          return refCount === 0 && cleanup
-            ? Fiber.interrupt(cleanup)
-            : Effect.void
-        })
+      let refCount = 0
+      let connector: PubSub.PubSub<Take.Take<A, E>> | null = null
+      let fiber: Fiber.RuntimeFiber<void> | null = null
+      const scope = yield* Effect.scope
+      const context = yield* Effect.context<R>()
+      return Effect.gen(function*() {
+        refCount++
+        connector ??= yield* (config.connector ?? PubSub.bounded<Take.Take<A, E>>(16))
+        fiber ??= yield* runIntoPubSub(self, connector).pipe(
```

Comment:
There is a potential race condition here, as the fiber could yield to the scheduler and another have another one sneak in.

Hard to reproduce and unlikely to happen, but possible.

Reply:
Is there a way to prevent it?

Reply:
You can use a semaphore or a SynchronizedRef to ensure only one access at a time.

Reply:
Is there a way to just declare an Effect as "unswitchable" or something like this? So the scheduler can not switch to another context until this effect is executed?

Reply:
Do you mean like this?

Reply:
Yes, a semaphore will prevent this.